Azure Arc-enabled servers Private Link integration (#915)
microsoft/azure_arc: docs/azure_arc_jumpstart/azure_arc_servers/day2/arc_privatelink/_index.md

---
type: docs
title: "Azure Private Link"
linkTitle: "Azure Private Link"
weight: 11
description: >
---

## Use Azure Private Link to securely connect networks to Azure Arc

The following README will guide you on how to use [Azure Private Link](https://docs.microsoft.com/en-us/azure/private-link/private-link-overview) to securely connect from an Azure Arc-enabled server to Azure using a VPN. [This feature](https://docs.microsoft.com/en-us/azure/azure-arc/servers/private-link-security) allows you to connect privately to Azure Arc without opening up public network access but rather using private endpoints over a VPN or ExpressRoute connection, ensuring that all traffic is being sent to Azure privately.

In this guide, you will emulate a hybrid environment connected to Azure over a VPN with hybrid resources that will be Arc-enabled, and Azure Private Link Scope will be used to connect over a private connection. To complete this process you deploy a single ARM template that will:

- Create two separate resource groups:

>   - "On-premises" resource group: will simulate a private on-premises environment with a Windows virtual machine. This VM does will not have a public IP address assigned to it so [Azure Bastion](https://docs.microsoft.com/en-us/azure/bastion/bastion-overview) is deployed to have administrative access to the operating system. The Windows virtual machine is an Azure Arc-enabled server by installing the Azure Arc-connected machine agent using Azure Private Link.

Reviewer: ```suggestion
  - "On-premises" resource group: will simulate a private on-premises environment with a Windows virtual machine. This VM will not have a public IP address assigned to it so [Azure Bastion](https://docs.microsoft.com/en-us/azure/bastion/bastion-overview) is deployed to have administrative access to the operating system. The Windows virtual machine is an Azure Arc-enabled server by installing the Azure Arc-connected machine agent using Azure Private Link.
```

  - "Azure" resource group: in this resource group, you will have the Azure Arc-enabled server and its Private Link Scope.

- Both resource groups have their own virtual networks and address spaces, and they are connected via Azure VPN gateways to set up a hybrid private connection.

>   ![Deployment Overview](./01.png)

Reviewer: Do we add a private link scope object in the diagram as well ?

Author: updated the image

>   > **Note: It is not expected for an Azure VM to be projected as an Azure Arc-enabled server. The below scenario is unsupported and should ONLY be used for demo and testing purposes.**

Reviewer: ```suggestion
  > **NOTE: It is not expected for an Azure VM to be projected as an Azure Arc-enabled server. The below scenario is unsupported and should ONLY be used for demo and testing purposes.**
```

>   > **Note: The below scenario assumes the on-premises VM has outbound internet connectivity for the deployment of the Azure Arc connected machine agent, for internet disconnected environments you will need to adjust the automation to retrieve the agent's software from locally accessible storage**

Reviewer: ```suggestion
  > **NOTE: The below scenario assumes the on-premises VM has outbound internet connectivity for the deployment of the Azure Arc connected machine agent, for internet disconnected environments you will need to adjust the automation to retrieve the agent's software from locally accessible storage**
```

## Prerequisites

- Clone the Azure Arc Jumpstart repository

    ```shell
    git clone https://github.com/microsoft/azure_arc.git
    ```

> - [Install or update Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). Azure CLI should be running version 2.7** or later. Use ```az --version``` to check your current installed version.

Reviewer: ```suggestion
- [Install or update Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). Azure CLI should be running version 2.25.0** or later. Use ```az --version``` to check your current installed version.
```

      ```

      You can monitor the registration process with the following commands:

      ```shell
      az provider show --namespace 'Microsoft.HybridCompute'
      az provider show --namespace 'Microsoft.GuestConfiguration'
      ```

## Automation Flow

For you to get familiar with the automation and deployment flow, below is an explanation.

1. User is editing the ARM template parameters file (1-time edit). These parameters values are being used throughout the deployment.

2. User deploys the ARM template at subscription level. The ARM template will create two resources groups with:

    - Azure resource group:
        - Azure Arc Private Link Scope
        - Azure Arc-enabled server
        - Azure Private Link Endpoint for the Azure Arc-enabled Server
        - Three Azure Private DNS zones
        - Azure VPN Gateway and its public IP address
        - Azure VNET

    - On-premises resource group:
        - Azure VNET
        - Azure Bastion
        - Azure VPN Gateway and its public IP address
        - Azure Windows Virtual Machine with a custom script extension that runs the **install_arc_agent.sh** script

>         > **Note: The [*install_arc_agent.sh*](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure/linux/arm_template/scripts/install_arc_agent.sh) shell script will enable the OS firewall and set up new rules for incoming and outgoing connections. By default all incoming and outgoing traffic will be allowed, except blocking Azure IMDS outbound traffic to the *169.254.169.254* remote address.**

Reviewer: ```suggestion
        > **NOTE: The [*install_arc_agent.sh*](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure/linux/arm_template/scripts/install_arc_agent.sh) shell script will enable the OS firewall and set up new rules for incoming and outgoing connections. By default all incoming and outgoing traffic will be allowed, except blocking Azure IMDS outbound traffic to the *169.254.169.254* remote address.**
```

- To deploy the ARM template, navigate to the local cloned [deployment folder](https://github.com/microsoft/azure_arc/tree/main/azure_arc_servers_jumpstart/privatelink) and run the below command, start deploying resources:

    ```shell
    az deployment sub create \
    --location <Azure Region Location> \
    --template-file <The *azuredeploy.json* template file location> \
    --parameters <The *azuredeploy.parameters.json* parameters file location>
    ```

    For example:

    ```shell
    az deployment sub create \
    --location eastus \
    --template-uri https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_servers_jumpstart/azure/privatelink/nesteddeploy.json \
    --parameters nesteddeploy.example.parameters.json
    ```

>      > **Note: The deployment may take around 45 minutes to complete**

Reviewer: ```suggestion
     > **NOTE: The deployment may take around 45 minutes to complete**
```

- Verify the resources are created on the Azure Portal for both resource groups:

    ![Resources created on Onpremises's resource group](./02.png)

    ![Resources created on Azure's resource group](./03.png)

## Windows Login & Post Deployment

> - Now that the Windows Server VM is created and the VPN connections are established, it is time to login to it. Using bastio RDP to the VM.

Reviewer: ```suggestion
- Now that the Windows Server VM is created and the VPN connections are established, it is time to RDP to it using Azure Bastion.
```

---
type: docs
title: "Azure Private Link"
linkTitle: "Azure Private Link"
weight: 11
description: >
---

## Use Azure Private Link to securrely connect networks to Azure Arc

> The following README will guide you on how to use [Azure Private Link](https://docs.microsoft.com/en-us/azure/private-link/private-link-overview) to securely connect from an Azure Arc-enabled server to Azure PaaS services. [This feature](https://docs.microsoft.com/en-us/azure/azure-arc/servers/private-link-security) not only allows you to link your Azure PaaS services to your virtual network using private endpoints but also enables you to connect your on-premises or multi-cloud resources with Azure Arc and ensure that all traffic is being send over a VPN or ExpressRoute connection.

Reviewer 2: ```suggestion
The following README will guide you on how to use [Azure Private Link](https://docs.microsoft.com/en-us/azure/private-link/private-link-overview) to securely connect from an Azure Arc-enabled server to Azure PaaS services. [This feature](https://docs.microsoft.com/en-us/azure/azure-arc/servers/private-link-security) not only allows you to link your Azure PaaS services to your virtual network using private endpoints but also enables you to connect your on-premises or multi-cloud resources with Azure Arc and ensure that all traffic is being sent over a VPN or ExpressRoute connection.
```

In this guide, you will emulate a hybrid environment connected to Azure over a VPN, hybrid resources will be Arc-enabled and Azure Private Link will be used to connect to an Azure PaaS service over a private connection. To complete this process you will:

> * Create two separate resource groups in different Azure regions:

Reviewer 2: ```suggestion
- Create two separate resource groups in different Azure regions:
```

Please standardized on dashes instead of * for now on

Author: got it

> In this guide, you will emulate a hybrid environment connected to Azure over a VPN, hybrid resources will be Arc-enabled and Azure Private Link will be used to connect to an Azure PaaS service over a private connection. To complete this process you will:
> 
> * Create two separate resource groups in different Azure regions:
> 
>   * On-premises resource group: will simulate a private on-premises environment with a Linux virtual machine, this VM does not have a public IP address assigned to it so Azure Bastion is deployed to have access to the operating system. The Linux virtual machine is Arc-enabled by installing the Azure Arc connected machine agent.
>   * Azure resource group: in this resource group you will deploy all Azure PaaS resources in this case Azure DNS and Azure SQL, to connect privately to them Azure Private Link is also deployed.

Reviewer 2: ```suggestion
In this guide, you will emulate a hybrid environment connected to Azure over a VPN, hybrid resources will be Arc-enabled and Azure Private Link will be used to connect to an Azure PaaS service over a private connection. To complete this process you will:

* Create two separate resource groups in different Azure regions:

  * "On-premises" resource group: will simulate a private on-premises environment with a Linux virtual machine, this VM does not have a public IP address assigned to it so Azure Bastion is deployed to have access to the operating system. The Linux virtual machine is Arc-enabled by installing the Azure Arc-connected machine agent.
  * Azure resource group: in this resource group you will deploy all Azure PaaS resources in this case Azure DNS and Azure SQL, to connect privately to them Azure Private Link is also deployed.
```

> * Both resource groups have their own virtual networks and address spaces, however they are connected via Azure VPN gateways to set up a hybrid private connection.

Reviewer 2: ```suggestion
* Both resource groups have their own virtual networks and address spaces, however, they are connected via Azure VPN gateways to set up a hybrid private connection.
```

>   ![Deployment Overview](./01.png)

Reviewer 2: Will refactor the diagram for you :-)

> Once everything is deployed you will be able to access the Azure SQL private IP address from the on-premises Linux machine, traffic will go over the VPN connection and them kept within the Azure VNET via Private Link to access the database service.

Reviewer 2: ```suggestion
Once everything is deployed, you will be able to access the Azure SQL private IP address from the "on-premises" Linux machine, traffic will go over the VPN connection and be kept within the Azure VNET via Private Link to access the database service.
```

## Prerequisites

* Clone the Azure Arc Jumpstart repository

    ```shell
    git clone https://github.com/microsoft/azure_arc.git
    ```

* [Install or update Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). Azure CLI should be running version 2.7** or later. Use ```az --version``` to check your current installed version.

* Create Azure service principal (SP)

    To connect a VM or bare-metal server to Azure Arc, Azure service principal assigned with the "Contributor" role is required. To create it, login to your Azure account run the below command (this can also be done in [Azure Cloud Shell](https://shell.azure.com/)).

    ```shell
    az login
    az ad sp create-for-rbac -n "<Unique SP Name>" --role contributor
    ```

    For example:

    ```shell
    az ad sp create-for-rbac -n "http://AzureArcServers" --role contributor
    ```

    Output should look like this:

    ```json
    {
    "appId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "displayName": "AzureArcServers",
    "name": "http://AzureArcServers",
    "password": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "tenant": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    }
    ```

> 

Reviewer 2: ```suggestion
- Azure Arc-enabled servers depends on the following Azure resource providers in your subscription in order to use this service. Registration is an asynchronous process, and registration may take approximately 10 minutes.

  - Microsoft.HybridCompute
  - Microsoft.GuestConfiguration

      ```shell
      az provider register --namespace 'Microsoft.HybridCompute'
      az provider register --namespace 'Microsoft.GuestConfiguration'
      ```

      You can monitor the registration process with the following commands:

      ```shell
      az provider show --namespace 'Microsoft.HybridCompute'
      az provider show --namespace 'Microsoft.GuestConfiguration'
      ```

```

  > **Note: It is optional but highly recommended to scope the SP to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest).**

## Automation Flow

For you to get familiar with the automation and deployment flow, below is an explanation.

1. User is editing the ARM templates parameters file (1-time edit). These parameters values are being used throughout the deployment. Two separate files are provided, one per resource group.

2. User creates two resource groups in two different regions:
    * Azure Resource Group
    * On-premises Resource Group

> 3. User runs the ARM template for the resources in the Azure Resource Group. The ARM template will create

Reviewer 2: ```suggestion
3. User deploys the ARM template for the resources in the Azure Resource Group. The ARM template will create
```

    * Azure SQL server and a SQL database

> 

Reviewer 2: no need for linespace between bulltets

    * Azure Private Link

    * Azure Private DNS zone

    * Azure VPN Gateway and its public IP address

    * Azure VNET

> 4. User runs the ARM template for the resources in the On-premises resource group. The ARM template will create: 

Reviewer 2: ```suggestion
4. User deploys the ARM template for the resources in the On-premises resource group. The ARM template will create: 
```

    * Azure VNET

> 

Reviewer 2: no need for linespace between bulltets

    * Azure Bastion

    * Azure VPN Gateway and its public IP address

>     * Azure Linux Virtual Machine with a custom script extension that will run the [*install_arc_agent.sh*](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure/linux/arm_template/scripts/install_arc_agent.sh) shell script to Arc-enable the Azure VM.

Reviewer 2: ```suggestion
    * Azure Linux Virtual Machine with a custom script extension that will run the [_install_arc_agent.sh_](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure/linux/arm_template/scripts/install_arc_agent.sh) shell script to Arc-enable the Azure VM.
```

Please standardized on underline for italic text from now on

>         > **Note: The [*install_arc_agent.sh*](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure/linux/arm_template/scripts/install_arc_agent.sh) shell script will enable the OS firewall and set up new rules for incoming and outgoing connections. By default all incoming and outgoing traffic will be allowed, except blocking Azure IMDS outbound traffic to the *169.254.169.254* remote address.**

Reviewer 2: ```suggestion
        > **Note: The [_install_arc_agent.sh_](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure/linux/arm_template/scripts/install_arc_agent.sh) shell script will enable the OS firewall and set up new rules for incoming and outgoing connections. By default all incoming and outgoing traffic will be allowed, except blocking Azure IMDS outbound traffic to the *169.254.169.254* remote address.**
```

> 5. User uses Azure Bastion to connect to Linux VM which will start the *install_arc_agent.sh* script execution and will onboard the VM to Azure Arc.

Reviewer 2: ```suggestion
5. User uses Azure Bastion to connect to Linux VM which will start the _install_arc_agent.sh_ script execution and will onboard the VM to Azure Arc.
```

6. User tests private connectivity to Azure SQL server from the Azure Linux Virtual Machine.

## Deployment

> As mentioned, this deployment will leverage ARM templates. You will deploy an ARM template per resource group that are responsible for creating all the Azure resources as well as the "on-premises" resources that will be onboarded to Azure Arc.

Reviewer 2: ```suggestion
As mentioned, this deployment will leverage ARM templates. You will deploy an ARM template per resource group that is responsible for creating all the Azure resources as well as the "on-premises" resources that will be onboarded to Azure Arc.
```

Can't we use a single template/param file using nested templates? makes the user experience much easier and less confusing IMO

Reviewer 2: Will continue reviewing following that decision

Author: I did try but the dependencies to do the VPN connection and since things are deployed in 2 different resources groups I could not make it work.

---
type: docs
title: "Azure Private Link"
linkTitle: "Azure Private Link"
weight: 11
description: >
---

> ## Use Azure Private Link to securrely connect networks to Azure Arc

Reviewer 2: ```suggestion
## Use Azure Private Link to securely connect networks to Azure Arc
```

The following README will guide you on how to use [Azure Private Link](https://docs.microsoft.com/en-us/azure/private-link/private-link-overview) to securely connect from an Azure Arc-enabled server to Azure PaaS services. [This feature](https://docs.microsoft.com/en-us/azure/azure-arc/servers/private-link-security) not only allows you to link your Azure PaaS services to your virtual network using private endpoints but also enables you to connect your on-premises or multi-cloud resources with Azure Arc and ensure that all traffic is being sent over a VPN or ExpressRoute connection.

> In this guide, you will emulate a hybrid environment connected to Azure over a VPN, hybrid resources will be Arc-enabled and Azure Private Link will be used to connect to an Azure PaaS service over a private connection. To complete this process you will:

Reviewer 2: ```suggestion
In this guide, you will emulate a hybrid environment connected to Azure over a VPN with hybrid resources that will be Arc-enabled, and Azure Private Link will be used to connect to an Azure PaaS service over a private connection. To complete this process you will:
```

- Create two separate resource groups in different Azure regions:

>   - "On-premises" resource group: will simulate a private on-premises environment with a Linux virtual machine, this VM does not have a public IP address assigned to it so Azure Bastion is deployed to have access to the operating system. The Linux virtual machine is Arc-enabled by installing the Azure Arc-connected machine agent.

Reviewer 2: ```suggestion
  - "On-premises" resource group: will simulate a private on-premises environment with a Linux virtual machine. This VM does will not have a public IP address assigned to it so [Azure Bastion](https://docs.microsoft.com/en-us/azure/bastion/bastion-overview) is deployed to have access to the operating system. The Linux virtual machine is an Azure Arc-enabled server by installing the Azure Arc-connected machine agent.
```

>   - Azure resource group: in this resource group you will deploy all Azure PaaS resources in this case Azure DNS and Azure SQL, to connect privately to them Azure Private Link is also deployed.

Reviewer 2: ```suggestion
  - Azure resource group: in this resource group, you will be deploying all Azure PaaS resources, in this case, [Azure DNS](https://docs.microsoft.com/en-us/azure/dns/dns-overview) and [Azure SQL](https://docs.microsoft.com/en-us/azure/azure-sql/database/sql-database-paas-overview). In order to establish a private connection to these Azure services, Azure Private Link will be deployed as well.
```

> - Both resource groups have their own virtual networks and address spaces, however they are connected via Azure VPN gateways to set up a hybrid private connection.

Reviewer 2: ```suggestion
- Both resource groups have their own virtual networks and address spaces, however, they are connected via Azure VPN gateways to set up a hybrid private connection.
```

  ![Deployment Overview](./01.png)

> Once everything is deployed, you will be able to access the Azure SQL private IP address from the "on-premises" Linux machine, traffic will go over the VPN connection and be kept within the Azure VNET via Private Link to access the database service.

Reviewer 2: ```suggestion
Once everything is deployed, you will be able to access the Azure SQL private IP address from the "on-premises" Linux machine while network traffic will go over the VPN connection and be kept within the Azure VNET via Private Link to access the database service.
```

    ```

    Output should look like this:

    ```json
    {
    "appId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "displayName": "AzureArcServers",
    "name": "http://AzureArcServers",
    "password": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "tenant": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    }
    ```

- Azure Arc-enabled servers depends on the following Azure resource providers in your subscription in order to use this service. Registration is an asynchronous process, and registration may take approximately 10 minutes.

  - Microsoft.HybridCompute
  - Microsoft.GuestConfiguration

      ```shell
      az provider register --namespace 'Microsoft.HybridCompute'
      az provider register --namespace 'Microsoft.GuestConfiguration'
      ```

      You can monitor the registration process with the following commands:

      ```shell
      az provider show --namespace 'Microsoft.HybridCompute'
      az provider show --namespace 'Microsoft.GuestConfiguration'
      ```

  > **Note: It is optional but highly recommended to scope the SP to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest).**

## Automation Flow

> For you to get familiar with the automation and deployment flow, below is an explanation.
> 
> 1. User is editing the ARM templates parameters file (1-time edit). These parameters values are being used throughout the deployment. Two separate files are provided, one per resource group.
> 
> 2. User creates two resource groups in two different regions:
>     - Azure Resource Group
>     - On-premises Resource Group
> 
> 3. User deploys the ARM template for the resources in the Azure Resource Group. The ARM template will create:

Reviewer 2: ```suggestion
For you to get familiar with the automation and deployment flow, below is an explanation.

1. User is editing the ARM templates parameters file (1-time edit). These parameter values are being used throughout the deployment. Two separate files are provided, one per resource group.

2. User creates two resource groups in two different regions:
    - Azure Resource Group
    - On-premises Resource Group

3. User deploys the ARM template for the resources in the Azure Resource Group. The ARM template will create:
```

    - Azure SQL server and a SQL database
    - Azure Private Link
    - Azure Private DNS zone
    - Azure VPN Gateway and its public IP address
    - Azure VNET

4. User deploys the ARM template for the resources in the On-premises resource group. The ARM template will create:

    - Azure VNET
    - Azure Bastion
    - Azure VPN Gateway and its public IP address
    - Azure Linux Virtual Machine with a custom script extension that will run the [_install_arc_agent.sh_](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure/linux/arm_template/scripts/install_arc_agent.sh) shell script to Arc-enable the Azure VM.

>         > **Note: The [_install_arc_agent.sh_](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure/linux/arm_template/scripts/install_arc_agent.sh) shell script will enable the OS firewall and set up new rules for incoming and outgoing connections. By default all incoming and outgoing traffic will be allowed, except blocking Azure IMDS outbound traffic to the *169.254.169.254* remote address.**

Reviewer 2: i would add the Azure VM disclaimer we have in the Azure VMs scenarios

Author: added

6. User tests private connectivity to Azure SQL server from the Azure Linux Virtual Machine.

## Deployment

As mentioned, this deployment will leverage ARM templates. You will deploy an ARM template per resource group that is responsible for creating all the Azure resources as well as the "on-premises" resources that will be onboarded to Azure Arc.

- Before deploying the ARM template, login to Azure using AZ CLI with the ```az login``` command.

- Create two resource groups in two different regions:

  - Azure Resource Group
  - On-premises Resource Group

    ```shell
    az group create --name <Name of the Azure resource group> --location <Azure region> --tags "Project=jumpstart_azure_arc_servers"
    az group create --name <Name of the On-premise resource group> --location <Azure region> --tags "Project=jumpstart_azure_arc_servers"
    ```

    For example:

    ```shell
    az group create --name Arc-Azure-PL-demo --location "westeurope" --tags "Project=jumpstart_azure_arc_servers"
    az group create --name Arc-Onpremise-PL-Demo --location "northeurope" --tags "Project=jumpstart_azure_arc_servers"
    ```

> - The deployment will use two ARM template parameters file. Before initiating the deployment, edit the [*azuredeploy.parameters.json*](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/privatelink/azuredeploy.parameters.json) file located in your local cloned repository folder and the edit the [*onpremisedeploy.parameters.json*](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/privatelink/onpremisedeploy.parameters.json). Example parameters files are located [here](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/privatelink/azuredeploy.example.parameters.json) and [here](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/privatelink/onpremisedeploy.example.parameters.json).

Reviewer 2: Can't we do this with a single ARM template?

Reviewer 3: We can, but the template will be quite large and difficult to troubleshoot/iterate on in the future. Is there a reason linked templates should be avoided?

Reviewer 3: To clarify, you only execute a single template here. The others are linked to the main template and deployed inline.

Reviewer 3: Looking through the code, I believe this sections in the Readme needs to be updated to reflect the single parameters file used for the nested deployment.

Reviewer 2: I didn't notice that so my bad. Although, The README is not doing a good job explaining it. 
Even though it's an implementation details, the README should make it clearer

Reviewer 3: Agreed, code is good, README needs updating to reflect actual deployment requirements.

Author: I'll review and make some changes to make it clearer

Author: something weird happened and I had not commited my changes for the index, should be fixed now, sorry

    --resource-group <Name of the Azure resource group> \
    --name <The name of this deployment> \
    --template-uri https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_servers_jumpstart/azure/privatelink/azuredeploy.json \
    --parameters <The *azuredeploy.parameters.json* parameters file location>
    ```

    For example:

    ```shell
    az deployment group create \
    --resource-group Arc-Azure-PL-demo \
    --name arcpldemo \
    --template-uri https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_servers_jumpstart/azure/privatelink/azuredeploy.json \
    --parameters azuredeploy.parameters.json
    ```

     > **Note: The deployment may take around 20-30 minutes to complete**

- Verify the resources are created on the Azure Portal.

    ![Resources created on Azure's resource group](./02.png)

- Once the deployment is done, create resources in the on-premises resource group. To deploy the ARM template, navigate to the local cloned [deployment folder](https://github.com/microsoft/azure_arc/tree/main/azure_arc_servers_jumpstart/privatelink) and run the below command:

    ```shell
    az deployment group create \

>     --resource-group <Name of the on-premise resource group> \

Reviewer 2: ```suggestion
    --resource-group <Name of the on-premises resource group> \
```

    --name <The name of this deployment> \
    --template-uri https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_servers_jumpstart/azure/privatelink/onpremise.json \
    --parameters <The *azuredeploy.parameters.json* parameters file location>
    ```

    For example:

    ```shell
    az deployment group create \
    --resource-group Arc-Onpremise-PL-Demo \
    --name arcpldemo \
    --template-uri https://raw.githubusercontent.com/microsoft/azure_arc/main/azure_arc_servers_jumpstart/azure/privatelink/onpremise.json \
    --parameters onpremisedeploy.parameters.json
    ```

     > **Note: The deployment may take around 20-30 minutes to complete**

- Verify the resources are created on the Azure Portal.

    ![Resources created on Onpremises's resource group](./03.png)

> - Once deployed, connect to the Linux virtual machine using Azure Bastion. Once you login, the *install_arc_agent.sh* script execution will start and will onboard the VM to Azure Arc.

Reviewer 2: ```suggestion
- Once deployed, connect to the Linux virtual machine using Azure Bastion. Once you log in, the _install_arc_agent.sh_ script execution will start and will onboard the VM to Azure Arc.
```

    ![Connect with Azure Bastion 01](./04.png)

    ![Connect with Azure Bastion 02](./05.png)

    ![Arc onboarding script 01](./06.png)

    ![Arc onboarding script 02](./07.png)

- The Azure Arc-enabled server will be shown in the Azure resource group as a new resource.

    ![Azure Arc-enabled server on Azure's resource group](./08.png)

## Connectivity test

> - Now that all resources are deployed in both resources groups, you can verify that there is a private connection from the Azure Arc-enabled server to the Azure SQL database, this connection will go through the VPN gateways that connect the two VNETs and the traffic will continue over the Microsoft's network to reach the PaaS service on its private IP address since Azure Private link has been enabled.

Reviewer 2: ```suggestion
- Now that all resources are deployed in both resources groups, you can verify that there is a private connection from the Azure Arc-enabled server to the Azure SQL database, this connection will go through the VPN gateways that connect the two VNETs and the network traffic will continue over the Microsoft's network to reach the PaaS service on its private IP address since Azure Private link has been enabled.
```

> - To perform the test you will need the database's private IP address, navigate to the Azure's resource group in the Azure Portal as described below:

Reviewer 2: ```suggestion
- To perform the test you will need the database's private IP address. Navigate to the resource group in the Azure Portal as described below:
```
